scripts/migration/courses-populate.ts

```diff
@@ -0,0 +1,171 @@
+    .then(jsonCourses => jsonCourses.map(jsonCourse => courseFieldFilter(jsonCourse)));
+  return courses;
+};
+
+/* Helper functions to clean and format class roster data */
+const cleanField = (value: string | null | undefined) =>
+  value?.replace(/\u00a0/g, ' ') || undefined;
+
+const courseFieldFilter = ({
+  subject,
+  crseId,
+  catalogNbr,
+  titleLong,
+  enrollGroups,
+  catalogWhenOffered,
+  catalogBreadth,
+  catalogDistr,
+  catalogComments,
+  catalogSatisfiesReq,
+  catalogCourseSubfield,
+  catalogAttribute,
+  acadCareer,
+  acadGroup,
+  description,
+  catalogPrereqCoreq,
+}: CourseFullDetail): CourseFullDetail => ({
+  subject: cleanField(subject) || '',
+  crseId,
+  catalogNbr: cleanField(catalogNbr) || '',
+  titleLong: cleanField(titleLong) || '',
+  enrollGroups,
+  catalogWhenOffered: cleanField(catalogWhenOffered),
+  catalogBreadth: cleanField(catalogBreadth),
+  catalogDistr: cleanField(catalogDistr),
+  catalogComments: cleanField(catalogComments),
+  catalogSatisfiesReq: cleanField(catalogSatisfiesReq),
+  catalogCourseSubfield: cleanField(catalogCourseSubfield),
+  catalogAttribute: cleanField(catalogAttribute),
+  acadCareer: cleanField(acadCareer) || '',
+  acadGroup: cleanField(acadGroup) || '',
+  description: cleanField(description) || '',
+  catalogPrereqCoreq: cleanField(catalogPrereqCoreq) || '',
+});
+
+/* 
+  Populates firebase with the retrieved course data, and meta data structures for future migration. 
+
+  availableRosterForCoursesCollection - contains a mapping from each course to a list of rosters that it is available in
+  crseToCatalogNbrCollection - contains a mapping from each course's crseId to its catalogNbr (which is a subject + code)
+  coursesCollection - contains all courses ever offered by the class roster API from FA14 - present. 
+*/
+const populateCourses = async (roster: string, subject: string, courses: CourseFullDetail[]) => {
+  courses.forEach(async course => {
+    /* Computing a mapping that maps a course to a list of semesters it is offered in */
+    const courseAndRostersDoc = await availableRostersForCourseCollection
+      .doc(`${subject} ${course.catalogNbr.toString()}`)
+      .get();
+    if (courseAndRostersDoc.exists) {
+      await availableRostersForCourseCollection
+        .doc(`${subject} ${course.catalogNbr.toString()}`)
+        .update({ rosters: FieldValue.arrayUnion(roster) });
+    } else {
+      await availableRostersForCourseCollection
+        .doc(`${subject} ${course.catalogNbr.toString()}`)
+        .set({ rosters: [roster] });
+    }
+    /* Mapping that maps from a course's crseId to their course code */
+    await crseIdToCatalogNbrCollection
+      .doc(course.crseId.toString())
+      .set({ catalogNbr: `${subject} ${course.catalogNbr}` });
+    /* Writing each course to the 'courses' collection */
+    await coursesCollection.doc(roster).collection(subject).doc(course.catalogNbr).set({ course });
+  });
+};
```

Comment:
At a glance this might not behave as expected; the promise returned by `populateCourses` will immediately resolve
```suggestion
const populateCourses = (roster: string, subject: string, courses: CourseFullDetail[]) =>
  Promise.all(courses.map(async course => {
    /* Computing a mapping that maps a course to a list of semesters it is offered in */
    const courseAndRostersDoc = await availableRostersForCourseCollection
      .doc(`${subject} ${course.catalogNbr.toString()}`)
      .get();
    if (courseAndRostersDoc.exists) {
      await availableRostersForCourseCollection
        .doc(`${subject} ${course.catalogNbr.toString()}`)
        .update({ rosters: FieldValue.arrayUnion(roster) });
    } else {
      await availableRostersForCourseCollection
        .doc(`${subject} ${course.catalogNbr.toString()}`)
        .set({ rosters: [roster] });
    }
    /* Mapping that maps from a course's crseId to their course code */
    await crseIdToCatalogNbrCollection
      .doc(course.crseId.toString())
      .set({ catalogNbr: `${subject} ${course.catalogNbr}` });
    /* Writing each course to the 'courses' collection */
    await coursesCollection.doc(roster).collection(subject).doc(course.catalogNbr).set({ course });
  }));
```

Reply:
Hmm. Could you explain a little bit why the `Promise.all() `is needed to wrap around the `courses.forEach()` loop? `courses` is just an array of data, not promises right? 

Outside of that, I do think the `async` in 118 should be removed, so thanks for that!

Reply:
`forEach` will not wait for the promise created for each iteration to resolve. My suggestion creates a promise for each course, and uses `Promise.all` to ensure each of those individual promises resolves before `populateCourses` resolves. Crucially, to do this, it also replaces `forEach` with `map`.

Reply:
Oh gotcha gotcha. you are referring to the async operations that are being applied to each course, and that `forEach` will not wait for those async operations to finish before moving onto the next course. That makes sense. Thanks!

Reply:
Yep. It's okay if the operations for each course finish in different orders (which they will now), but we want all of them to finish before `populateCourses` resolves.